### DISCUSSION TITLE
[@alova/wormhole] 希望提供集中式管理API方案

### DISCUSSION THREAD

**Discussion by Lnncoco** [alovajs/alova#613] [General]
比如我有一个URL拼装参数后跳转的场景，调用Apis.xxx()传入参数后，希望不是直接请求，而是获取组装后的URL地址，真正去请求是打开新页面或者由其他组件响应。

**Comment by Lnncoco** [comment]
同一个问题，生成的包相当于api的管理器，那么其他零散的api希望归集在一起管理。但目前设计似乎倾向于外部api单独管理。
比如，某些原因下，可能要对后端接口进行固定的二次封装，或者多个接口集合一起封装成一个调用接口，这类接口也希望从Apis里能够直接导出使用，好像没找到方法。
如果对接外部系统，提供了第三方无swagger情况下，这类api似乎只能另外新建文件自行管理？无法很好的与Apis或者在生成后的apis目录下集中式管理。

**Reply by czhlin** [reply]
目前只是通过opnapi文件按照它的规则生成的全局Apis，目前确实没有提供可以组合api成一个api的方法，也没有非openapi文件挂载到Apis的方法。 @JOU-amjs 帮忙看一下需要支持这些特性吗😂

**Reply by JOU-amjs** [reply]
对了，对一些接口进行二次封装是指怎么样的封装？以及，组合多个接口具体是指怎么样的组合？

可否举一些具体点例子，用伪代码表示下也可以

**Reply by Lnncoco** [reply]
类似自行添加一个新的接口方法，里面可以调用到其他已经生成的swagger方法，去做请求处理。
举例的场景可能不太正规，比如两个获取list的方法数据，因为需求变动，暂时只需要前端将两个接口数据合并，后期再统一成一个。这时前端在api管理层先处理了，后期只需要替换api地址即可。
也就是api管理的问题。

**Reply by JOU-amjs** [reply]
我理解了，我个人认为这个需求有点特殊，自己手动添加一个函数处理可能会更好，我们处理这种情况顶多添加apis自定义管理的支持其实是收效比较小的

**Reply by Lnncoco** [reply]
> 我理解了，我个人认为这个需求有点特殊，自己手动添加一个函数处理可能会更好，我们处理这种情况顶多添加apis自定义管理的支持其实是收效比较小的

是的，比较特殊。其实想表达的还是api集中式管理，这种场景有了集中式管理（添加自定义的能力），再加上内部能调用到其他swagger控制变量，就能够支持非常多场景了。

**Comment by MeetinaXD** [comment]
@czhlin 需要帮忙看看 🤔

**Comment by JOU-amjs** [comment]
@Lnncoco 目前是没有拼装URL的功能，你可以直接用`URLSearchParams`和`URL`类可以快速实现。
```js
const params = new URLSearchParams(method.config.params || {});
const url = new URL(method.url + '?'  + params.toString(), method.baseURL);
```

第二是关于将零散api归集到Apis中一起管理的想法，这个很不错，我们思考下怎么来设计它会更好。

**Reply by Lnncoco** [reply]
我尝试 Apis.xxx.xxx 里调出url地址，但没成功。希望通过Apis里可以获取到目标地址，万一swagger更新后，也能自动更新这个，而不用写死

**Reply by MeetinaXD** [reply]
你需要获取到method对象才可以访问url属性。

``` js
const method = Apis.general.methodName()
```

**Comment by Lnncoco** [comment]
补充一个场景，Swagger 不支持 WebSocket 的展示，这块往往需要手动来配置地址，如果有集中化管理，并且能从 Apis 中直接调出 url 来应用，会更方便管理
